### PR TITLE
Refactor specs to share setup among instance + class contexts

### LINKED ISSUE #129: Refactor test setup to share among instance and class contexts

### PR DESCRIPTION
**Pull request by ms-ati**
Closes #129

More steps:
  * Refactor specs: use comments to break up sections of specs
  * Refactor specs: move instance method tests into contexts
  * Refactor specs: extract shared examples for #memo_wise
  * Extract definition of instance methods for testing
  * Extract to a support file: definition of methods for testing
  * Fix common test setup among instance/class/class-scope contexts
  * Run shared examples for class << self context
  * Simplify test method setup by reusing instance for class << self

### CODE REVIEW

File: spec/memo_wise_spec.rb [hunk 1]
```diff
@@ -3,1071 +3,917 @@
...
+            protected_memowise_method_counter + 1
+          "protected_memowise_method"
+        end
+        protected :protected_memowise_method
+        memo_wise :protected_memowise_method
 
-      def with_positional_and_keyword_args_counter
-        @with_positional_and_keyword_args_counter || 0
-      end
+        def self.class_positional_args_counter
```
Comment (JacobEvelyn) [inline spec/memo_wise_spec.rb:37]: Comments explaining why these methods are here would be really helpful.

File: spec/memo_wise_spec.rb [hunk 2]
```diff
@@ -3,1071 +3,917 @@
...
-      def true_method_counter
-        @true_method_counter || 0
-      end
+        def instance_with_keyword_args_counter
+          @instance_with_keyword_args_counter || 0
+        end
 
-      def nil_method_counter
-        @nil_method_counter || 0
+        def with_keyword_args(a:, b:) # rubocop:disable Naming/MethodParameterName
```
Comment (JacobEvelyn) [inline spec/memo_wise_spec.rb:63]: Same here.

File: spec/memo_wise_spec.rb (hunk 1 above)
Comment (ms-ati) [reply spec/memo_wise_spec.rb:40]: Comments added!

File: spec/memo_wise_spec.rb (hunk 2 above)
Comment (ms-ati) [reply spec/memo_wise_spec.rb:77]: Comments added!

**Review by JacobEvelyn — APPROVED** [review]

**Pull request merged**